File: ios/PriceFinder/PriceFinder/PriceFinding/WalmartScraper.swift
//
//  WalmartScraper.swift
//  PriceFinder
//
//  Created by jacob stimes on 8/6/16.
//  Copyright © 2016 stimes.enterprises. All rights reserved.
//

import Foundation

/// Queries Walmart API
class WalmartScraper : Scraper, ScraperProtocol {
    
    let url = "http://api.walmartlabs.com/v1/search?";
    
    let apiKeyParam = "apiKey";
    
    let queryParam = "query";
    
    let apiKey = Config.WalmartApiKey;
    
    
    init() {
        super.init(site: "Walmart");
    }
    
    func search(keywords: String) -> [Item] {
        self.keywords = keywords;
        
        let s = NSCharacterSet.URLQueryAllowedCharacterSet().mutableCopy();
        s.addCharactersInString("+&");
        var query = queryParam + "=" + self.keywords!.stringByAddingPercentEncodingWithAllowedCharacters(s as! NSCharacterSet)! + "&";
        
        query = query + apiKeyParam + "=" + apiKey.stringByAddingPercentEncodingForRFC3986()!;
        
        let queryUrl = self.url + query;
        
        let json = self.parseIntoJSON(self.getJSON(queryUrl));
        return getItemsFromJSON(json);
    }
    
    func getItemsFromJSON(results: JSON) -> [Item] {
        var items = [Item]();
        
        let itemsNode = "items";
        
        let itemsJsonArray = results[itemsNode] as! [JSON]
        for itemJson in itemsJsonArray {
            let name = itemJson["name"] as! String;
            let salePrice = itemJson["salePrice"] as! NSNumber;
            let link = itemJson["productUrl"] as! String;
            
            let thumbnailLink = itemJson["thumbnailImage"] as! String;
            
            items.append(Item(title: name, price: Double(salePrice), vendor: site, link: link, thumbnailLink: thumbnailLink));
            
        }
        
        return items;
    }
    
}

File: android/PriceFinder/app/src/main/java/enterprises/stimes/pricefinding/scraping/WalmartScraper.java
package enterprises.stimes.pricefinding.scraping;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.ArrayList;

import org.json.JSONException;
import org.json.JSONObject;

public class WalmartScraper extends Scraper {
	
	private String searchUrl = "http://api.walmartlabs.com/v1/search?";

	private String apiKeyParam = "apiKey";
	private String queryParam = "query";
	private String apiKey = Config.walmartApiKey;
	
	public WalmartScraper(){
		super("Walmart");
	}
	
	public ArrayList<ScrapingItem> search(String queryString) {
		this.keywords = queryString;
		try {
			String query = String.format("%s=%s&%s=%s", 
				 apiKeyParam,
			     URLEncoder.encode(this.apiKey, charset), 
			     queryParam,
			     URLEncoder.encode(this.keywords, charset));
			
			String queryUrl = searchUrl + query;
			System.out.println(queryUrl);
			JSONObject json = ScrapeUtils.readJsonFromUrl(queryUrl);
			//System.out.println("NumItems: " + Long.toString((long)json.get("numItems")));
			return super.getItemsFromSearch(json);
		} 
		catch (UnsupportedEncodingException e1) {
			e1.printStackTrace();
			return new ArrayList<ScrapingItem>();
		}
		catch(JSONException e){
			e.printStackTrace();
			return new ArrayList<>();
		}
	}

}
